awscli/customizations/s3/results.py

```diff
@@ -21,21 +21,31 @@
 
 LOGGER = logging.getLogger(__name__)
 
-QueuedResult = namedtuple(
-    'QueuedResult',
-    ['transfer_type', 'src', 'dest', 'total_transfer_size']
-)
-ProgressResult = namedtuple(
-    'ProgressResult',
-    ['transfer_type', 'src', 'dest', 'bytes_transferred',
-     'total_transfer_size']
-)
-SuccessResult = namedtuple(
-    'SuccessResult', ['transfer_type', 'src', 'dest']
-)
-FailureResult = namedtuple(
-    'FailureResult', ['transfer_type', 'src', 'dest', 'exception']
-)
+
+def _create_new_result_cls(name, extra_fields=None):
+    # Creates a new namedtuple class that subclasses from BaseResult for the
+    # benefit of filtering by type and ensuring particular base attrs.
+
+    # NOTE: _fields is a public attribute that has an underscore to avoid
+    # naming collisions for namedtuples:
+    # https://docs.python.org/2/library/collections.html#collections.somenamedtuple._fields
+    fields = list(BaseResult._fields)
+    if extra_fields:
+        fields += extra_fields
+    return type(name, (namedtuple(name, fields), BaseResult), {})
+
+
+BaseResult = namedtuple('BaseResult', ['transfer_type', 'src', 'dest'])
+
+QueuedResult = _create_new_result_cls('QueuedResult', ['total_transfer_size'])
```

Comment:
It's a little weird to have `_create_new_result_cls` implicitly depend on `BaseResult` which is defined above.  If the `BaseResult` line is moved below a `_create_new_result_cls` the function will fail.  What do you think about passing `BaseResult` to `_create_new_result_cls` to make this more explicit?